### DISCUSSION TITLE
Maximum enthalpy for a given pressure (and understanding if that is to be trusted)

### DISCUSSION THREAD

**Discussion by fwitte** [CoolProp/CoolProp#2443] [Q&A]
Hi all,

disclaimer: I am really happy, that I do not have to deal with the implementation of fluid property functions myself due to your work on CoolProp! That also means, whenever possible I did not dive too deep in related topics. However, I would like to understand something I came across recently:

I am working with cyclopentane calling `AbstractState` functions via `HmassP_inputs` (because my software, https://github.com/oemof/tespy/, is based on ph as variables). In order to not invoke `ValueErrors` when interfacing with CoolProp, I make sure to check whether my inputs are within the minimum and maximum pressure as well as minimum and maximum enthalpy (which I derive from actual pressure of a state point and the minimum and maximum temperature provided in the trivial outputs of a fluid on the CoolProp side). I am happy with that implementation, because it works quite well in terms of limiting the mathematics in a Newton solver to stay within physical boundaries and not crash CoolProp.

However, I somehow came across, that I can call with enthalpy inputs that correspond to a temperature higher than stated in the docs (550 K, according to http://www.coolprop.org/fluid_properties/fluids/Cyclopentane.html, and also calling the `trivial_keyed_output(CP.iT_max)`). I investigated a little bit and found, that the actual temperature limit for which I can retrieve a result seems to be at ~552 °C:

```python
import CoolProp as CP


fluid = "Cyclopentane"
state = CP.AbstractState("HEOS", fluid)

state.update(CP.QT_INPUTS, 1, 200 + 273.15)
p_evap = state.p()

T_max = state.trivial_keyed_output(CP.iT_max)
state.update(CP.PT_INPUTS, p_evap, T_max)
print(f"Enthalpy at maximum temperature {state.hmass()}")

T = T_max
while True:
    state.update(CP.PT_INPUTS, p_evap, T)
    h = state.hmass()
    try:
        state.update(CP.HmassP_INPUTS, h, p_evap)
        assert round(state.T()) == T  # assert, that functions are still closed here
    except ValueError:
        break

    T += 1

print(f"Limit according to 'T_max': {T_max}")
print(f"First failed temperature (= actual limit?): {T}")
print(f"Corresponding enthalpy: {state.hmass()}")
```

which outputs:

```
Enthalpy at maximum temperature 775065.0687740949
Limit according to 'T_max': 550.0
First failed temperature: 826.0
Corresponding enthalpy: 1576153.1910588525
```

I would like to understand, what the reason for that is. And, I guess, that the implementation is valid + tested until the 550 K stated in the fluid json/docs, above that you can retrieve results, but they are not necessarily correct. Is my guess correct? I could not really find any information in the docs on this topic.

Thanks a lot, have a sunny weekend!!

Francesco

### Versions

**CoolProp Version:** 6.6.0
**Operating System and Version:** Windows 10
**Access Method:** see code snippet

**Comment by fwitte** [comment]
I tested it for the other fluids as well, (no surprise) it is always the same ratio with respect to `T_max`, i.e. 1.5:

```python
import CoolProp as CP


for fluid in CP.__fluids__:
    state = CP.AbstractState("HEOS", fluid)

    T = state.trivial_keyed_output(CP.iT_max * 0.8)
    state.update(CP.QT_INPUTS, 1, T)
    p_evap = state.p()

    T_max = state.trivial_keyed_output(CP.iT_max)
    state.update(CP.PT_INPUTS, p_evap, T_max)
    print(f"Enthalpy at maximum temperature {state.hmass()}")

    T = T_max
    while True:
        state.update(CP.PT_INPUTS, p_evap, T)
        h = state.hmass()
        try:
            state.update(CP.HmassP_INPUTS, h, p_evap)
            assert round(state.T()) == T  # assert, that functions are still closed here
        except ValueError:
            break

        T += 1

    print(f"Last successfull evaluation: {fluid} -> {T / T_max} * T_max")
```

That come from some location inside CoolProp, maybe here: https://github.com/CoolProp/CoolProp/blob/a988270be2dc736bef3cda06830bcc5eaec1debf/src/Backends/Helmholtz/FlashRoutines.cpp#L1174?

**Reply by henningjp** [reply]
The constant factor of 1.5 is telling.

I know that REFPROP allows (and has a flag to disallow) extrapolation up to 1.5*Tmax and 2.0*Pmax as there is very little error in doing so.  Perhaps CoolProp allows the same thing?  I did a little digging but could not verify this assertion.

**Reply by ibell** [reply]
Yes the factor *should* be 1.5Tmax
